### PR TITLE
Use warp level primitives for accumulate

### PR DESCRIPTION
**Pull request by Ellipse0934**
Replace Belloch Scan with warp level scan.

Mostly works right now. Still a little refinement lacking
Benchmark:
length = 2<sup>27</sup>, Float32, addition
Previous: 0.070033128 seconds
Current:   0.032186161 seconds

Small changes like now `init` is performed in the aggregate_partial_sums. Let me know if there are any issues with this. Will mark for review when completed.

### CODE REVIEW

**Comment by Ellipse0934** [comment]
Changes from previous implementation:
- Exclusive scan now works
- There was a bug with `init` which has been fixed.
- Each element is mapped to 1 thread, block size fixed to 256

There is a comment ([here](https://github.com/JuliaGPU/CUDA.jl/blob/master/src/accumulate.jl#L13-L14))
```
# - the ND case is quite a bit slower than the 1D case (not using Cartesian indices,
#   before 35fcbde1f2987023229034370b0c9091e18c4137). optimize or special-case?
```

I couldn't get that branch to run on the first try but the current performance is identical to a C++ version of `scan` I wrote for a linear array so maybe there is no longer an issue. There is a launch configuration bug (`blocks_others`) where we are launching way more blocks than needed, maybe correcting this resolved it. My question is whether it is necessary to investigate this or can it be skipped ?

There is bug in the new and current versions. Since CUDA doesn't support grid dims where y or z axis > 2<sup>16</sup>(65536). The kernel will fail to execute when `length(Rpre)` or `length(Rpost)` exceeds that. Maybe address this in a subsequent PR.

Why are we not exporting `scan!` ?

**Comment by maleadt** [comment]
Nice, thanks! I'll give this a proper review as soon as I have some time.



> I couldn't get that branch to run on the first try

The hash is from before the CUDA.jl merge, so it effectively points to what was CuArrays.jl.



> Why are we not exporting `scan!` ?

We don't export additional array operations unless there's something GPU specific about them, only implement Base stuff.

bors try

**Comment by Ellipse0934** [comment]
bors try

**Comment by Ellipse0934** [comment]
> The hash is from before the CUDA.jl merge, so it effectively points to what was CuArrays.jl.

Yes, I had tried on CuArrays.jl but it didn't work on the 1<sup>st</sup> go. It probably needs other relevant packages to be near the same commit, my Julia was also 1.4, so...

Tests did not pass on my system earlier due type instability which has been fixed now. But, probably does need a review.

**Comment by maleadt** [comment]
bors try

**Comment by Ellipse0934** [comment]
^ It actually passed but there are failures for other tests unrelated to the PR.

**Comment by maleadt** [comment]
I've restarted those tests.

File: src/accumulate.jl
Comment (maleadt) [inline src/accumulate.jl]: `wid, lane = fldmod1(threadIdx().x, warpsize())`

File: src/accumulate.jl
Comment (maleadt) [inline src/accumulate.jl]: could you keep the variables the same, for a reduced diff?

File: src/accumulate.jl
Comment (maleadt) [inline src/accumulate.jl]: 1-based indices are more usual in Julia

File: src/accumulate.jl
Comment (maleadt) [inline src/accumulate.jl]: you should use the occupancy API for this

File: src/accumulate.jl
Comment (maleadt) [inline src/accumulate.jl]: This breaks on older devices that don't have `shfl`. See our mapreduce implementation which has a fallback for those devices.

**Review by maleadt — CHANGES_REQUESTED** [review]
Quick review, looking good! Requires some changes though, maybe you can pick some techniques from our `mapreduce` implementation which is obviously pretty similar.

File: src/accumulate.jl
Comment (Ellipse0934) [reply src/accumulate.jl]: What's the minimum compute capability do we plan to support ? `SHFL` intrinsics are supported from 3.0 which seemed like a decent starting point to me. I am guessing support for any prior compute capability is only for people on old computers with integrated laptops where speed is not a priority.

Comment (maleadt) [reply src/accumulate.jl]: Maybe it's time to bump it to 3.0, but then we should either warn about that during `__init__`, or error out when calling this function. I'll see about that in another PR.

**Comment by Ellipse0934** [comment]
@maleadt 
> threads_per_block = 256;

> you should use the occupancy API for this

There is a reason I have done this instead of the occupancy API has to do with performance. I am imagining that in the future functions like `mapreduce` and `scan` each have a table for each graphics card and depending on the card the "correct"(most performant) launch configuration is chosen. The occupancy API only considers the kernel and the data-type (and memory to avoid spillage which won't happen here), not the data-size, it's a thumb rule heuristic rather than an absolute answer for optimization.  Running the occupancy API for my `partial_scan` results in 1024 as the number of threads. Performance wise, comparing 256(current) to the suggested(1024) we get:
(256)
```julia
julia> using CUDA
[ Info: Precompiling CUDA [052768ef-5323-5732-b1bb-66c8b64840ba]

julia> sizes = [5, 10, 15, 20, 25, 26];

julia> function times(sizes)
           bench = [];
           for x in sizes
               len = 1 << x
               a = CuArray(1.:len)
               b = similar(a);
               CUDA.scan!(+, b, a);
              CUDA.synchronize();
              t = Base.@elapsed begin
                   CUDA.scan!(+, b, a);
                   CUDA.synchronize();
               end
               push!(bench, t);
               CUDA.unsafe_free!(a);
               CUDA.unsafe_free!(b);
           end
           return bench
           end
times (generic function with 1 method)

julia> sizes = [5, 10, 15, 20, 25, 26];

julia> times(sizes)
6-element Array{Any,1}:
 0.000206172
 0.000119258
 0.000104623
 0.000361647
 0.010636605
 0.020962764

julia> times(sizes)
6-element Array{Any,1}:
 0.000330406
 0.000534187
 0.000516692
 0.000336635
 0.00965961
 0.019635128

julia> times(sizes)
6-element Array{Any,1}:
 0.00033227
 0.000530968
 0.000276619
 0.000359589
 0.010299
 0.020932708
```

(1024)
```julia
julia> times(sizes)
6-element Array{Any,1}:
 0.000206852
 7.2228e-5
 0.000127156
 0.000480146
 0.014437776
 0.027361384

julia> times(sizes)
6-element Array{Any,1}:
 0.000351906
 0.000305722
 0.000558829
 0.000429986
 0.013004558
 0.025484215

julia> times(sizes)
6-element Array{Any,1}:
 0.000349574
 0.000304593
 0.000557104
 0.000481173
 0.014450518
 0.027304227
```
I know this isn't a great benchmark especially for the small sized arrays where the launch is dominating time but it does show that at least for large sized arrays (2<sup>25/26</sup>) the difference is significant enough. With 256 elements per thread(and 2<sup>26</sup>) being a max 0.020 and 1024 per thread being 0.025, 0.027 and 0.025. As array sizes get larger this trend will probably continue as this is a memory intensive kernel rather than compute intensive, we want to issue a fetch and then switch to another warp to hide latency of fetch(400 cycles if I remember correctly). I do plan to work on such infrastructure in the near future.

With a large number of packages which will use CUDA's functions like mapreduce, broadcast, scan,.etc I think it's worth it to despite the cost of more work and uglier code for performance. We should take every 5% performance we get. 

Should I still put the launch configurator here and do my work in future PR's ?

File: src/accumulate.jl
Comment (Ellipse0934) [reply src/accumulate.jl]: To put it in perspective with the year of release:
3.0 : Kepler Architecture : 2012 release
5.0 : Maxwell Architecture: 2014 release

Have a look [here](https://en.wikipedia.org/wiki/CUDA#Version_features_and_specifications), something like 3.0 or 3.5 as the minimum version makes sense to me. I doubt there are users serious about running on 3.0 compute capability.

File: src/accumulate.jl
Comment (Ellipse0934) [reply src/accumulate.jl]: I found the variable names `tid`, `wid`, `bid`, `wid` more pneumonic to their purpose here rather than `i` and `j`. We can change other places in the package to keep this uniform. Your call obviously but I think its better my way and makes the code easier to understand.

File: src/accumulate.jl
Comment (vchuravy) [reply src/accumulate.jl]: There are likely people out there who are using an old K40/K50 those were very popular cards.

Comment (Ellipse0934) [reply src/accumulate.jl]: K40 is 3.5 . But how many people are going to be using anything made before that ?

**Comment by maleadt** [comment]
Good to know. Did you specify shared memory to the configurator API to make sure its decisions were accurate?

> Should I still put the launch configurator here and do my work in future PR's ?

Not necessarily, but I'm worried that we'll run into situations where 256 threads is going to be too much. With higher-order, generically typed functions like this, it's easy enough to inline a very complex user lambda or work with types whose operations are much more complex (DualNumbers, Intervals, etc) and result in significant shared memory or register pressure.

**Comment by Ellipse0934** [comment]
Yes, I did put in shared memory but it is obviously very low here. Maybe, for this PR I can put in `threads = max(min(256, OCC_launch.threads), 32)` to be conservative.

**Comment by Ellipse0934** [comment]
@maleadt 
Maybe, the approach for map, reduce, scan should be that after generating the `ptx` we should see how many registers are used
every function's header is something like
```
        .reg .b64       %SP;
        .reg .b64       %SPL;
        .reg .pred      %p<4>;
        .reg .b16       %rs<3>;
```
Which gives an idea of how many registers are going to get used by a thread. Maybe the number of registers change slightly when the sass is generated. Or maybe we can get the register count from the `sass` code. In any case I'd like to know if this is a decent strategy and I will investigate it. For this PR it is better if I shift it to the launch configurator (Tomorrow morning).

Other than that everything from the initial review has been addressed except the variable naming scheme: i.e. 
>   i = (blockIdx().x-1) * blockDim().x + threadIdx().x

to

>   tid = (blockIdx().x-1) * blockDim().x + threadIdx().x

and similarly using tid, bid, wid and lid, which I still prefer

**Comment by Ellipse0934** [comment]
Turns out I messed up earlier with the launch configurator and it is suggesting 640 threads /  block instead of 1024. 
(640)
```julia
julia> times(sizes)
6-element Array{Any,1}:
 0.0002721
 0.000462318
 0.000451718
 0.000539891
 0.014663179
 0.02928899

julia> times(sizes)
6-element Array{Any,1}:
 0.000377855
 0.000611228
 0.000594773
 0.001034141
 0.013335861
 0.026649408

julia> times(sizes)
6-element Array{Any,1}:
 0.000405838
 0.00061428
 0.000600149
 0.000871435
 0.012780693
 0.02621747
```

Which shows that there is still some room for improvement. And my previous suggestion of looking at the ptx can be ignored, there is a function [cudaFuncGetAttributes](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1g19b1696533069c03f646e2ce2beacc00) which will shows the register usage per thread. Present inside CUDA.jl [here](https://github.com/JuliaGPU/CUDA.jl/blob/463a41295bfede5125c584e6be9c51a4b9074e12/src/device/cuda/libcudadevrt.jl#L12). This does not work yet but seems straightforward.

**Comment by maleadt** [comment]
> there is a function [cudaFuncGetAttributes](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EXECUTION.html#group__CUDART__EXECUTION_1g19b1696533069c03f646e2ce2beacc00) which will shows the register usage per thread. Present inside CUDA.jl [here](https://github.com/JuliaGPU/CUDA.jl/blob/463a41295bfede5125c584e6be9c51a4b9074e12/src/device/cuda/libcudadevrt.jl#L12).

That's device side, host-side you can compile a kernel using `cufunction` (the low-level API `@cuda` expands to) and call `registers` on it. But now that we have the occupancy API you rarely need that anymore.

File: src/accumulate.jl
Comment (maleadt) [reply src/accumulate.jl]: OK, fair enough, I'm happy to bump the required capability. but `shfl` is only supported for certain types, `(Bool, Int32, Int64, Float32, Float64, ComplexF32, ComplexF64)`, so we need a fallback anyhow.

**Comment by maleadt** [comment]
Some non-technical suggestions:
- better use a branch that isn't `master` on your fork, makes it easier to checkout by other people
- never commit merge conflicts. if you accidentally to so, rebase them out
- better just don't merge, but rebase on top of master

File: src/accumulate.jl
Comment (maleadt) [reply src/accumulate.jl]: `tid` and `bid` make me think of thread ID and block ID, which it doesn't mean here, right? Both are an index across the entire iteration space. So I've reverted that for now to keep the diff smaller and make it easier to review.

**Comment by maleadt** [comment]
Why the hell did GitHub close this... (That's what I meant with not using the `master` branch)

Anyhow, please open a new PR if you want. I squashed your branch because it was beyond saving. Please stick a little closer to the original code style though (comment style, whitespace, etc), needless changes like that are annoying to review.

I also don't really get a bunch of the changes, e.g. getting rid of `init=nothing`, not passing `init` to the scan kernels anymore. Are you sure that doesn't break anything?

I'll have a look at adding some more test coverage and benchmarks to the master branch so we can evaluate this properly :slightly_smiling_face:

**Pull request closed without merging**
